File: twitter_client.py
import os
import tweepy


def get_client():
    return tweepy.Client(
        bearer_token=os.getenv("TWITTER_BEARER_TOKEN"),
        consumer_key=os.getenv("TWITTER_API_KEY"),
        consumer_secret=os.getenv("TWITTER_API_SECRET"),
        access_token=os.getenv("TWITTER_ACCESS_TOKEN"),
        access_token_secret=os.getenv("TWITTER_ACCESS_TOKEN_SECRET"),
        wait_on_rate_limit=True,
    )


def get_my_user_id(client: tweepy.Client) -> str:
    resp = client.get_me(user_auth=True)
    if not resp.data:
        raise ValueError("Не удалось определить твой аккаунт — проверь TWITTER_ACCESS_TOKEN")
    return str(resp.data.id)


def get_target_user_id(client: tweepy.Client, username: str) -> str:
    resp = client.get_user(username=username, user_auth=False)
    if not resp.data:
        raise ValueError(f"Аккаунт @{username} не найден")
    return str(resp.data.id)


def get_liked_tweet_ids(client: tweepy.Client, my_user_id: str) -> set[str]:
    resp = client.get_liked_tweets(
        id=my_user_id,
        max_results=100,
        tweet_fields=["id"],
        user_auth=True,
    )
    if not resp.data:
        return set()
    return {str(t.id) for t in resp.data}


def get_last_3_posts(username: str) -> tuple[list[dict], list[dict]]:
    """Returns (unliked, all_tweets) for the last 3 posts of username."""
    client = get_client()
    my_user_id = get_my_user_id(client)
    target_user_id = get_target_user_id(client, username)
    liked_ids = get_liked_tweet_ids(client, my_user_id)

    resp = client.get_users_tweets(
        id=target_user_id,
        max_results=5,
        tweet_fields=["id", "text", "created_at"],
        exclude=["retweets", "replies"],
    )
    if not resp.data:
        return [], []

    all_tweets, unliked = [], []
    for tweet in resp.data[:3]:
        t = {
            "id": str(tweet.id),
            "text": tweet.text,
            "url": f"https://x.com/{username}/status/{tweet.id}",
            "liked": str(tweet.id) in liked_ids,
        }
        all_tweets.append(t)
        if not t["liked"]:
            unliked.append(t)

    return unliked, all_tweets

File: twitterClient.ts
import "dotenv/config";
import { TwitterApi } from "twitter-api-v2";

export interface Tweet {
  id: string;
  text: string;
  url: string;
  liked: boolean;
}

function userClient() {
  return new TwitterApi({
    appKey: process.env.TWITTER_API_KEY!,
    appSecret: process.env.TWITTER_API_SECRET!,
    accessToken: process.env.TWITTER_ACCESS_TOKEN!,
    accessSecret: process.env.TWITTER_ACCESS_TOKEN_SECRET!,
  });
}

function appClient() {
  return new TwitterApi(process.env.TWITTER_BEARER_TOKEN!).readOnly;
}

async function getMyUserId(): Promise<string> {
  const me = await userClient().v2.me();
  if (!me.data?.id) throw new Error("Could not get your Twitter user ID");
  return me.data.id;
}

async function getLikedIds(myUserId: string): Promise<Set<string>> {
  const result = await userClient().v2.userLikedTweets(myUserId, { max_results: 100 });
  return new Set((result.data?.data ?? []).map((t) => t.id));
}

export async function validateUsername(username: string): Promise<string> {
  const result = await appClient().v2.userByUsername(username);
  if (!result.data?.id) throw new Error(`NOT_FOUND`);
  return result.data.id;
}

export async function getLast3Posts(username: string): Promise<{ unliked: Tweet[]; all: Tweet[] }> {
  const myUserId = await getMyUserId();
  const targetUserId = await validateUsername(username);
  const likedIds = await getLikedIds(myUserId);

  const timeline = await appClient().v2.userTimeline(targetUserId, {
    max_results: 5,
    exclude: ["retweets", "replies"],
    "tweet.fields": ["id", "text", "created_at"],
  });

  const tweets = (timeline.data?.data ?? []).slice(0, 3);
  const all: Tweet[] = [];
  const unliked: Tweet[] = [];

  for (const tweet of tweets) {
    const t: Tweet = {
      id: tweet.id,
      text: tweet.text,
      url: `https://x.com/${username}/status/${tweet.id}`,
      liked: likedIds.has(tweet.id),
    };
    all.push(t);
    if (!t.liked) unliked.push(t);
  }

  return { unliked, all };
}
